include/caffe/util/db.hpp

```diff
@@ -0,0 +1,188 @@
+#ifndef CAFFE_UTIL_DB_HPP
+#define CAFFE_UTIL_DB_HPP
+
+#include <string>
+
+#include "leveldb/db.h"
+#include "leveldb/write_batch.h"
+#include "lmdb.h"
+
+#include "caffe/common.hpp"
+#include "caffe/proto/caffe.pb.h"
+
+namespace caffe { namespace db {
+
+enum Mode { READ, WRITE, NEW };
+
+class Cursor {
+ public:
+  Cursor() { }
+  virtual ~Cursor() { }
+  virtual void SeekToFirst() = 0;
+  virtual void Next() = 0;
+  virtual string key() = 0;
+  virtual string value() = 0;
+  virtual bool valid() = 0;
+
+  DISABLE_COPY_AND_ASSIGN(Cursor);
+};
+
+class Transaction {
+ public:
+  Transaction() { }
+  virtual ~Transaction() { }
+  virtual void Put(const string& key, const string& value) = 0;
+  virtual void Commit() = 0;
+
+  DISABLE_COPY_AND_ASSIGN(Transaction);
+};
+
+class DB {
+ public:
+  DB() { }
+  virtual ~DB() { }
+  virtual void Open(const string& source, Mode mode) = 0;
+  virtual void Close() = 0;
+  virtual Cursor* NewCursor() = 0;
+  virtual Transaction* NewTransaction() = 0;
+
+  DISABLE_COPY_AND_ASSIGN(DB);
+};
+
+class LevelDBCursor : public Cursor {
+ public:
+  explicit LevelDBCursor(leveldb::Iterator* iter)
+    : iter_(iter) { SeekToFirst(); }
+  ~LevelDBCursor() { delete iter_; }
+  virtual void SeekToFirst() { iter_->SeekToFirst(); }
+  virtual void Next() { iter_->Next(); }
+  virtual string key() { return iter_->key().ToString(); }
+  virtual string value() { return iter_->value().ToString(); }
+  virtual bool valid() { return iter_->Valid(); }
+
+ private:
+  leveldb::Iterator* iter_;
+};
+
+class LevelDBTransaction : public Transaction {
+ public:
+  explicit LevelDBTransaction(leveldb::DB* db) : db_(db) { CHECK_NOTNULL(db_); }
+  virtual void Put(const string& key, const string& value) {
+    batch_.Put(key, value);
+  }
+  virtual void Commit() {
+    leveldb::Status status = db_->Write(leveldb::WriteOptions(), &batch_);
+    CHECK(status.ok()) << "Failed to write batch to leveldb "
+                       << std::endl << status.ToString();
+  }
+
+ private:
+  leveldb::DB* db_;
+  leveldb::WriteBatch batch_;
+
+  DISABLE_COPY_AND_ASSIGN(LevelDBTransaction);
+};
+
+class LevelDB : public DB {
+ public:
+  LevelDB() : db_(NULL) { }
+  virtual ~LevelDB() { Close(); }
+  virtual void Open(const string& source, Mode mode);
+  virtual void Close() {
+    if (db_ != NULL) {
+      delete db_;
+      db_ = NULL;
+    }
+  }
+  virtual LevelDBCursor* NewCursor() {
+    return new LevelDBCursor(db_->NewIterator(leveldb::ReadOptions()));
+  }
+  virtual LevelDBTransaction* NewTransaction() {
+    return new LevelDBTransaction(db_);
+  }
+
+ private:
+  leveldb::DB* db_;
+};
+
+inline void MDB_CHECK(int mdb_status) {
+  CHECK_EQ(mdb_status, MDB_SUCCESS) << mdb_strerror(mdb_status);
+}
+
+class LMDBCursor : public Cursor {
+ public:
+  explicit LMDBCursor(MDB_txn* mdb_txn, MDB_cursor* mdb_cursor)
+    : mdb_txn_(mdb_txn), mdb_cursor_(mdb_cursor), valid_(true) {
```

Comment:
I was sloppy about this, assuming a nonempty database... will do it properly now.